### PR TITLE
Enable multiline_parameter & multiline_arguments rules in SwiftLint

### PR DESCRIPTION
**Pull request by TheAdamBorek**
I've started to enabling the set of rules described in proposal. However, those rules affects more than our StyleGuide says so I want to first look if majority of us wants to have this rule enabled with all its pros & cons

### CODE REVIEW

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+The documentation with samples what particular rule does can be found [here](https://github.com/realm/SwiftLint/blob/master/Rules.md#multiline-arguments).
+
+## Impact on existing codebase
+After adding rules to our codebase, Xcode reports about 2500+ violations of them. That would need to be solved before marging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there's around ~900 issues to be fixed manually.
+
+In my opinion the impact on the codebase will be very positive, however it will also add few changes which are not defined in our style guide. I think it's acceptable price to make it possible to resolve nits quicker which would decrease a time when a PR is open.
+
+However, the given set of rules goes a step further. `multiline_arguments_brackets` rule makes it possible achieve what we want, which is, having a trailing bracket at a function call in the new line `)`. However, this rule **always** requires to put trailing bracket `)` in a newline if the function takes more than just one line to be invoked. Even if it's only one argument.
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: > Even if it's only one argument.

Can you add an example of what it means on practice?

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Few example what pieces of the code trigger the rule and how can this be fixed:
+
+ 1. **Trailing closure argument**:
+```swift
+AppsFlyerTracker.shared()?.continue(userActivity, restorationHandler: { restoring in
+        restorationHandler(restoring?.compactMap({ $0 as? UIUserActivityRestoring }))
+})
+```
+The fix:
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: 😍

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+ 2. **Inlined array argument**:
+```swift
+NSLayoutConstraint.activate([
+    imageView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -viewModel.viewSpec.secondaryLogoDistanceFromBottom),
+    imageView.centerXAnchor.constraint(equalTo: view.centerXAnchor)
+])
+```
+The fix:
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: 😍

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+NSLayoutConstraint.activate(
+    [
+        imageView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -viewModel.viewSpec.secondaryLogoDistanceFromBottom),
+        imageView.centerXAnchor.constraint(equalTo: view.centerXAnchor)
+    ]
+)
+```
+or:
+```swift
+NSLayoutConstraint.activate(
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: 😢

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+ 3. **Multiline inner argument**:
+```swift
+return accessible.access(InsuranceDetailsService.saveInsuranceDetails(
+    membershipId: membershipId,
+    selectedProvider: selectedProvider,
+    patientId: patientId
+))
+```
+The fix:
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: 😍

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+)
+```
+
+More examples of changes that would need to be added can be found in [the sample PR](https://github.com/Babylonpartners/babylon-ios/pull/7246/files) I've created against develop. It's not ready to be merged. It's a result of running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) & couple of manual changes.
+ 
+## Alternatives considered
+
+There are few alternatives:
+
+- drop the one of those four rules (the `multiline_arguments_brackets`). However dropping this rule will allow to have such a function call, which is against our style guide:
```
Comment (ilyapuchka) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: Looking at examples of suggested rules the only ensure that there is no mix on parameters defined on separate lines or in one line, but they still allow first parameter to be defined on the line of function call or declaration:

```swift
func some(paramA: A,
          paramB: B)
```

Personally I'd prefer us to stick with one variant, which is

```swift
func some(
    paramA: A,
    paramB: B
)
```

**Review by ilyapuchka — APPROVED** [review]
Love it! Lately I find myself more and more often putting parameters on separate lines, sometimes even when they are short, sometimes depending on the surrounding content, so that it is nicely boxed and nothing sticks out.

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 1 above)
Comment (TheAdamBorek) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: @ilyapuchka there are examples beloew ;)

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 4 above)
Comment (TheAdamBorek) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: yes, this is also the one I don't like the most

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 1 above)
Comment (ilyapuchka) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: In this case it might collide with code style for arrays, so I was looking for example that would be free of other rules. Will this:

```swift
func foo(bar: Bar)
```

always trigger violation and will require to change it to this?

```swift
func foo(
  bar: Bar
)
```

I'm not necessary against this edge case.

Comment (TheAdamBorek) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: @ilyapuchka When I read the proposal again I think I know what has mislead you to ask that question ;)
Here is updated version of one sentence :

> Few examples what pieces of the code trigger the rule and how can this be fixed. Keep in mind this is a list of changes which are not defined in the Function Definition or Function Call sections of the StyleGuide.

TL;DR All examples I've shown in the proposal are changes which don't come from the `Function Definition` or `Function Call` rules from our Style Guide. Those are examples of changes which are also forced on us to bring to our Style Guide if we decide to merge the proposal.

**Comment by TheAdamBorek** [comment]
just to be clear. One liner functions with mutliple arguments are still possible:

```swift
foo(a: A, b: B, c: C)
```
doesn't violate SwiftLint rules until all the arguments are in the same line

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 1 above)
Comment (ilyapuchka) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: So what you are trying to say with this note then is that 

```swift
foo(bar: Bar())
```

or 

```swift
foo(
  bar: Bar()
)
``` 

will be fine, but 

```swift
foo(
   bar: Bar())
```

will be not, right?

**Review by diegopetrucci — APPROVED** [review]
I am not quite sure if I can approve a proposal in which I am the review manager, if not please dismiss my vote.

This is a good change and I'm all for it, good job Adam.

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  spline: splines,
+  adjustmentFactor: 1.3,
+  translateConstant: 2,
+  comment: "normalize the display"
+)
+```
+To reduce our time spent on fixing such nits we can use SwiftLint to warns us about such issues.
+
+## Proposed solution
+Turn on following SwiftLint rules:
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
Turn on the following SwiftLint rules:
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+```
+- multiline_arguments
+- multiline_arguments_brackets
+- multiline_parameters
+- multiline_parameters_brackets
+```
+The documentation with samples what particular rule does can be found [here](https://github.com/realm/SwiftLint/blob/master/Rules.md#multiline-arguments).
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
The documentation of these rules, with examples, can be found [here](https://github.com/realm/SwiftLint/blob/master/Rules.md#multiline-arguments).
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+## Impact on existing codebase
+After adding rules to our codebase, Xcode reports about 2500+ violations of them. That would need to be solved before marging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there's around ~900 issues to be fixed manually.
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
After adding these rules to our codebase, Xcode reports about 2500+ violations. That would need to be solved before marging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there are around ~900 issues to be fixed manually.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 1 above)
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
However, the given set of rules goes a step further. `multiline_arguments_brackets` rule makes it possible achieve what we want — which is, having a trailing bracket at a function call in the new line `)`. However, this rule **always** requires to put the trailing bracket `)` in a newline if the function takes more than just one line to be invoked. Even if it's only one argument.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+In my opinion the impact on the codebase will be very positive, however it will also add few changes which are not defined in our style guide. I think it's acceptable price to make it possible to resolve nits quicker which would decrease a time when a PR is open.
+
+However, the given set of rules goes a step further. `multiline_arguments_brackets` rule makes it possible achieve what we want, which is, having a trailing bracket at a function call in the new line `)`. However, this rule **always** requires to put trailing bracket `)` in a newline if the function takes more than just one line to be invoked. Even if it's only one argument.
+
+Few examples what pieces of the code trigger the rule and how can this be fixed. Keep in mind this is a list of changes which are not defined in the `Function Definition` or `Function Call` sections of the [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations).
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
These are a few examples of pieces of code that trigger the rule and how can this be fixed. Keep in mind that this is a list of changes which are not defined in the `Function Definition` or `Function Call` sections of the [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations), and so the style guide will need to be updated accordingly.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+There are few alternatives:
+
+- drop the one of those four rules (the `multiline_arguments_brackets`). However dropping this rule will allow to have such a function call, which is against our style guide:
+```swift
+makeAutoFillFormAction(with: formProperties,
+    genders: genders,
+    continueAction: continueAction)
+```
+
+- write custom rule to swiftlint which would satisfy our needs. We could consider creating a PR to the SwiftLint. This however will be more time consuming to do.
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
- Write a custom rule to swiftlint which would satisfy our needs. We could consider creating a PR to the SwiftLint repo. This however will be more time consuming to do.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+)
+```
+
+More examples of changes that would need to be added can be found in [the sample PR](https://github.com/Babylonpartners/babylon-ios/pull/7246/files) I've created against develop. It's not ready to be merged. It's a result of running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) & couple of manual changes.
+ 
+## Alternatives considered
+
+There are few alternatives:
+
+- drop the one of those four rules (the `multiline_arguments_brackets`). However dropping this rule will allow to have such a function call, which is against our style guide:
```
Comment (diegopetrucci) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
- Drop the one of those four rules — the `multiline_arguments_brackets`. However, doing so will allow to have such a function call, which is against our style guide:
```

**Review by diegopetrucci — APPROVED** [review]

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md (hunk 1 above)
Comment (TheAdamBorek) [reply Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]: I've updated the proposal. I hope now it's more on point ;)

My problem is the following. What we want to achieve is to have function parameters/arguments defined in separate lines as well to have brackets in separate lines:

```swift
func reticulateSplines(
  spline: [Double],
  adjustmentFactor: Double,
  translateConstant: Int,
  comment: String
) -> Bool {
  // reticulate code goes here
}

let success = reticulateSplines(
  spline: splines,
  adjustmentFactor: 1.3,
  translateConstant: 2,
  comment: "normalize the display"
)
```

SwiftLint has a rule for that but the rule also require changes in places which are not defined in our style guide. Examples `Trailing closure argument` `Inlined array argument` `Multiline inner argument ` show when **additional** changes would be required.

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
+# Enable multiline rules in SwiftLint
+
+* Author(s): Adam Borek
+* Review Manager: Diego Petrucci
+
+## Introduction
+The motivation to create this proposal to not waste the time to fix violations of a SwfitLint's rule which we maybe wouldn't like to have, considering all pros & cons.
+
+## Motivation
+In our [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations) we have a section about multiline arguments & parameters. Multiline function declaration or function calls should have their brackets in separate lines.To reduce our time spent on fixing such nits we can use SwiftLint to warns us about such issues.
```
Comment (Ben-Henshall) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
In our [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations) we have a section about multiline arguments & parameters. Multiline function declaration or function calls should have their brackets in separate lines. To reduce our time spent on fixing such nits we can use SwiftLint to warns us about such issues.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
+# Enable multiline rules in SwiftLint
+
+* Author(s): Adam Borek
+* Review Manager: Diego Petrucci
+
+## Introduction
+The motivation to create this proposal to not waste the time to fix violations of a SwfitLint's rule which we maybe wouldn't like to have, considering all pros & cons.
```
Comment (Ben-Henshall) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
The motivation to create this proposal is to not waste time fixing violations of a SwfitLint rule which we maybe wouldn't like to have, considering all pros & cons.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+let success = reticulateSplines(
+  spline: splines,
+  adjustmentFactor: 1.3,
+  translateConstant: 2,
+  comment: "normalize the display"
+)
+```
+This is all about having parameters/arguments in seprate lines when we define/call multiple parameters/arguments functions.
```
Comment (Ben-Henshall) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
This is all about having parameters/arguments in separate lines when we define/call multiple parameters/arguments functions.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+This time, the following examples show what else triggers the rule and how can this be fixed. Keep in mind that this is a list of changes which are not defined in the `Function Definition` or `Function Call` sections of the [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations), and so the style guide will need to be updated accordingly. I would say sometimes the code before a fix is more readable than after SwiftLint's fix.
```
Comment (Ben-Henshall) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
This time, the following examples show what else triggers the rule and how this can be fixed. Keep in mind that this is a list of changes which are not defined in the `Function Definition` or `Function Call` sections of the [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations), and so the style guide will need to be updated accordingly. I would say sometimes the code before a fix is more readable than after SwiftLint's fix.
```

**Review by Ben-Henshall — APPROVED** [review]
It would be nice to accommodate to suggest multiline params/arguments for one liners over 120 character length at some point, but this is a great improvement for now 👍

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+```
+- multiline_arguments
+- multiline_arguments_brackets
+- multiline_parameters
+- multiline_parameters_brackets
+```
+The documentation of these rules, with examples, can be found [here](https://github.com/realm/SwiftLint/blob/master/Rules.md#multiline-arguments).
+
+## Impact on existing codebase
+After adding these rules to our codebase, Xcode reports about 2500+ violations. That would need to be solved before marging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there are around ~900 issues to be fixed manually.
```
Comment (inamiy) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
After adding these rules to our codebase, Xcode reports about 2500+ violations. That would need to be solved before merging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there are around ~900 issues to be fixed manually.
```

File: Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        membershipId: membershipId,
+        selectedProvider: selectedProvider,
+        patientId: patientId
+    )
+)
+```
+
+More examples of changes that would need to be added can be found in [the sample PR](https://github.com/Babylonpartners/babylon-ios/pull/7246/files) I've created against develop. It's not ready to be merged. It's a result of running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) & couple of manual changes.
+
+## The question to be asked when reviwing this proposal
```
Comment (inamiy) [inline Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md]:
```suggestion
## The question to be asked when reviewing this proposal
```

**Review by inamiy — APPROVED** [review]

**Comment by TheAdamBorek** [comment]
I got 4 approvals so I think I can merge this without creating a meeting :)
I will work on the swiftlint's rules later this or next week
